src/table/io/FitsSchemaInputMapper.cc

```diff
@@ -0,0 +1,619 @@
+    // on column number, flag bit, and name (ttype).  This allows us to insert fields into the
+    // schema in the correct order, regardless of which order they appear in the
+    // FITS header.
+    typedef boost::multi_index_container<
+        FitsSchemaItem,
+        boost::multi_index::indexed_by<
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::column>
+            >,
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::bit>
+            >,
+            boost::multi_index::hashed_unique<
+                boost::multi_index::member<FitsSchemaItem,std::string,&FitsSchemaItem::ttype>
+            >,
+            boost::multi_index::sequenced<>
+        >
+    > InputContainer;
+
+    // Typedefs for the special functors used to set data members.
+    typedef SetFitsSchemaString<&FitsSchemaItem::ttype> SetTTYPE;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tform> SetTFORM;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tccls> SetTCCLS;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tunit> SetTUNIT;
+    typedef SetFitsSchemaString<&FitsSchemaItem::doc> SetDoc;
+
+    // Typedefs for the different indices.
+    typedef InputContainer::nth_index<0>::type ByColumn;
+    typedef InputContainer::nth_index<1>::type ByBit;
+    typedef InputContainer::nth_index<2>::type ByName;
+    typedef InputContainer::nth_index<3>::type AsList;
+
+    // Getters for the different indices.
+    ByColumn & byColumn() { return inputs.get<0>(); }
+    ByBit & byBit() { return inputs.get<1>(); }
+    ByName & byName() { return inputs.get<2>(); }
+    AsList & asList() { return inputs.get<3>(); }
+
+    Impl() : version(0), flagColumn(0), archiveHdu(-1) {}
+
+    int version;
+    int flagColumn;
+    int archiveHdu;
+    Schema schema;
+    std::vector<std::unique_ptr<FitsColumnReader>> readers;
+    std::vector<Key<Flag>> flagKeys;
+    std::unique_ptr<bool[]> flagWorkspace;
+    PTR(io::InputArchive) archive;
+    InputContainer inputs;
+};
+
+FitsSchemaInputMapper::FitsSchemaInputMapper(daf::base::PropertyList & metadata, bool stripMetadata) :
+    _impl(boost::make_shared<Impl>())
+{
+    // Set the table version.  If AFW_TABLE_VERSION tag exists, use that
+    // If not, set to 0 if it has an AFW_TYPE, Schema default otherwise (DM-590)
+    if (!metadata.exists("AFW_TYPE")) {
+        _impl->version = lsst::afw::table::Schema::VERSION;
+    }
+    _impl->version = metadata.get("AFW_TABLE_VERSION", _impl->version);
+    if (stripMetadata) {
+        metadata.remove("AFW_TABLE_VERSION");
+    }
+
+    // Find a key that indicates an Archive stored on other HDUs
+    _impl->archiveHdu = metadata.get("AR_HDU", -1);
+    if (stripMetadata && _impl->archiveHdu > 0) {
+        metadata.remove("AR_HDU");
+    }
+
+    // read aliases stored in the new, expected way
```

Comment:
I'll clarify the code comment.  There is no "old way" for aliases, just an old way for slots, which predate aliases.

Your concern about how official the FITS format is is a timely one; I recently had a discussion with Gregory about how the SUI team can access some of the metadata in our FITS files that goes beyond what's standardized by FITS itself, and since they're using entirely different languages they'd like to avoid using afw to do the reading.  I definitely don't want to start formalizing our on-disk formats beyond what's in the FITS standard to the extent that we'll be unable to change those formats in the future, but I do anticipate we'll want to start documenting things like aliases for external users in the future - with a large "use at your own risk" disclaimer.

I'm going to treat that as a separate issue, though.